iceoryx_hoofs/test/moduletests/test_posix_mutex.cpp

```diff
@@ -66,126 +65,207 @@ class Mutex_test : public Test
-    lockedMutexBlocks(this, sutRecursive);
+    lockedMutexBlocks(this, *sutRecursive);
+}
+
+#ifndef _WIN32
+TEST_F(Mutex_test, MutexWithDeadlockDetectionsFailsOnDeadlock)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "feb07935-674d-4ebc-abaa-66664751719a");
+    iox::cxx::optional<iox::posix::mutex> sut;
+    ASSERT_FALSE(
+        iox::posix::MutexBuilder().mutexType(iox::posix::MutexType::WITH_DEADLOCK_DETECTION).create(sut).has_error());
+    EXPECT_FALSE(sut->lock().has_error());
+    auto result = sut->lock();
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(iox::posix::MutexLockError::DEADLOCK_CONDITION));
+
+    EXPECT_FALSE(sut->unlock().has_error());
+}
+#endif
+
+TEST_F(Mutex_test, MutexWithDeadlockDetectionsFailsWhenSameThreadTriesToUnlockItTwice)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "062e411e-a5d3-4759-9faf-db6f4129d395");
+    iox::cxx::optional<iox::posix::mutex> sut;
+    ASSERT_FALSE(
+        iox::posix::MutexBuilder().mutexType(iox::posix::MutexType::WITH_DEADLOCK_DETECTION).create(sut).has_error());
+    EXPECT_FALSE(sut->lock().has_error());
+    EXPECT_FALSE(sut->unlock().has_error());
+
+    auto result = sut->unlock();
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(iox::posix::MutexUnlockError::NOT_OWNED_BY_THREAD));
+}
+
+TEST_F(Mutex_test, MutexWithDeadlockDetectionsFailsWhenAnotherThreadTriesToUnlock)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "4dcea981-2259-48c6-bf27-7839ad9013b4");
+    iox::cxx::optional<iox::posix::mutex> sut;
+    ASSERT_FALSE(
+        iox::posix::MutexBuilder().mutexType(iox::posix::MutexType::WITH_DEADLOCK_DETECTION).create(sut).has_error());
+    EXPECT_FALSE(sut->lock().has_error());
+
+    std::thread t([&] {
+        auto result = sut->unlock();
+        ASSERT_TRUE(result.has_error());
+        EXPECT_THAT(result.get_error(), Eq(iox::posix::MutexUnlockError::NOT_OWNED_BY_THREAD));
+    });
+    t.join();
+    EXPECT_FALSE(sut->unlock().has_error());
+}
+
+#if !defined(__APPLE__) && !defined(_WIN32)
+TEST_F(Mutex_test,
+       MutexWithOnReleaseWhenLockedBehaviorUnlocksLockedMutexWhenThreadTerminatesAndSetsItIntoInconsistentState)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "4da7b1fb-23f1-421c-acf3-2a3d9e26b1a1");
+    iox::cxx::optional<iox::posix::mutex> sut;
+    ASSERT_FALSE(iox::posix::MutexBuilder()
+                     .threadTerminationBehavior(iox::posix::MutexThreadTerminationBehavior::RELEASE_WHEN_LOCKED)
+                     .create(sut)
+                     .has_error());
+
+    std::thread t([&] { EXPECT_FALSE(sut->lock().has_error()); });
+    t.join();
+
+    auto result = sut->try_lock();
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(),
+                iox::posix::MutexTryLockError::LOCK_ACQUIRED_BUT_HAS_INCONSISTENT_STATE_SINCE_OWNER_DIED);
+    sut->make_consistent();
+    EXPECT_FALSE(sut->unlock().has_error());
 }
 
+#if !defined(__FreeBSD__)
```

Comment:
I'm surprised that the test doesn't work on FreeBSD.

Reply:
Here the freebsd posix implementation seems to slightly diverge from the linux/qnx one. But I think with the native mutex implementation: https://www.freebsd.org/cgi/man.cgi?query=mutex&sektion=9 one could reimplement the behavior. A task for the far future.